### ISSUE TITLE
interface extends Array need initial index key

### ISSUE THREAD

**Issue by aaabinbin** [microsoft/TypeScript#35726]
I want to achieve a fixed size array 
```
type FixedLengthArray<T extends any, L extends number> = Array<T> & {
  0: T,
  length: L;
}

type ArrayValue = 'a' | 'b'

type Foo = FixedLengthArray<ArrayValue, 2>;
const foo1: Foo = ['a', 'b'];
```
But when I remove `0: T`, it throws a error, and Array has defined `[n: number]: T`, what does '0' means.
```
Type '("a" | "b")[]' is not assignable to type 'FixedLengthArray<ArrayValue, 2>'.
  Type '("a" | "b")[]' is not assignable to type '{ length: 2; }'.
    Types of property 'length' are incompatible.
      Type 'number' is not assignable to type '2'.
```

**Comment by dragomirtitian** [comment]
Why not use a tuple type ? 

```ts
type Foo = [ArrayValue, ArrayValue]
const foo1: Foo = ['a', 'b'];
```
[Playground Link](https://www.typescriptlang.org/play/#code/C4TwDgpgBAggTnAhiAaogNgV2gXigckXygB8CAjfAWAChRIoAxAe2ajwG14lUNsAaWAmRosEALq0AxswB2AZ2BQAZqwCMALiat2UDoXyD8lcQG4gA)

My guess as to the error is that typescript uses the presence of a numeric `0` property as a marker to for inferring tuples from array literals. When inferring a tuple the `length` property will be a numeric literal type (`2` in this case). If the property is missing you end up with an array type from the array literal and thus the `length` property will just be the regular `number` instead of `2`

**Comment by aaabinbin** [comment]
@dragomirtitian thx~ for your answer, I have found the related issue [here](https://github.com/mstn/fixed-size-array/issues/1#issuecomment-56657055)
